Update "The Monad type class" section to reflect modern Haskell (#160)
learnyouahaskell/learnyouahaskell.github.io: source_md/a-fistful-of-monads.md

```{.haskell:hs}
class Applicative m => Monad m where
    return :: a -> m a

>     return = pure

Reviewer: Removal of return from the main text is on my wish / to-do list.

Reviewer: I plan to take this on myself later.

    (>>=) :: m a -> (a -> m b) -> m b

    (>>) :: m a -> m b -> m b
    x >> y = x >>= \_ -> y
```

![this is you on monads](assets/images/a-fistful-of-monads/kid.png){.right width=363 height=451}

Let's start with the first line.

> It says `class Applicative m => Monad m where`, which means that if we want to create an instance of `Monad` for some type, we must also have an instance of `Applicative` for that type.

Reviewer: The reader has seen this kind of thing before. It is explained in chapter 8, and encountered twice (assuming #146) in the wild in the previous chapter, so I think this addition can be safely dropped.

Reviewer: Thread continued at https://github.com/learnyouahaskell/learnyouahaskell.github.io/pull/160#discussion_r3193433736.

Reviewer 2: hopefully resolved by applying the suggestion in the thread you referenced

> 

Reviewer: ```suggestion
```
(New thread because GH doesn't support expanding line selection.)

Reviewer: Or explicitly reference back:

```suggestion
In order to make a `Monad` instance, we first need an `Applicative` instance, just like `Applicative` itself needs `Functor`.

```

Next up, we have `>>`.
We won't pay too much attention to it for now because it comes with a default implementation and we pretty much never implement it when making `Monad` instances.

Now that we know what the `Monad` type class looks like, let's take a look at how `Maybe` is an instance of `Monad`!

```{.haskell:hs}
instance Monad Maybe where
    Nothing >>= f = Nothing
    Just x >>= f  = f x
```

Both `return` and `(>>)` have default implementations, so we usually omit them in instances.

> 

Reviewer 2: ```suggestion
```

Next up, we have `>>`.
We won't pay too much attention to it for now because it comes with a default implementation and we pretty much never implement it when making `Monad` instances.

Now that we know what the `Monad` type class looks like, let's take a look at how `Maybe` is an instance of `Monad`!

```{.haskell:hs}
instance Monad Maybe where
    Nothing >>= f = Nothing
    Just x >>= f  = f x
```

> Both `return` and `(>>)` have _default implementations_, so we can omit them in instances.

Reviewer: Adding this doesn't seem necessary. Default implementations have been discussed and come up several times prior, including two sentences up.

Reviewer: Thread continued at https://github.com/learnyouahaskell/learnyouahaskell.github.io/pull/160#discussion_r3193439807.

> `return` is the same as `pure`; it wraps a value in `Just`.

Reviewer: ```suggestion
```
(New thread because GH doesn't support expanding line selection.)

Reviewer 2: ```suggestion
Both `return` and `(>>)` have default implementations, so we usually omit them in instances.
```

I still think it doesn't hurt to remind but I made this reminder shorter. Hopefully, it's all right now.

If the matching falls through all the patterns for a given function, an error is thrown and our program crashes.
On the other hand, failed pattern matching in `let` expressions results in an error being produced right away, because the mechanism of falling through patterns isn't present in `let` expressions.

> When pattern matching fails in a `do` expression, the `fail` function is called.

Reviewer: This used to be true, and it isn't exactly false nowadays, but. 😐

Programs with refutable patterns in `do` blocks no longer just crash. Instead, they fail to compile: GHC requires such patterns to come with `MonadFail` instances.

Reviewer: I'll probably attempt to address this myself later.

> It's part of the `MonadFail` type class and it enables failed pattern matching to result in a failure in the context of the current monad, instead of making our program crash.

Reviewer 2: ```suggestion
It's a part of the `MonadFail` type class, and it enables failed pattern matching to result in a failure in the context of the current monad instead of making our program crash.
```

> It's a part of the `MonadFail` type class, and it enables failed pattern matching to result in a failure in the context of the current monad instead of making our program crash.

Reviewer: Very minor, but I saw your edit.

I think «it's part of» is more idiomatic here. An extra «a» suggests that there are other parts, and calls attention to this.

Reviewer: ```suggestion
It's part of the `MonadFail` type class, and it enables failed pattern matching to result in a failure in the context of the current monad instead of making our program crash.
```

We apply `show` to `x` to turn our number into a string and then we check if the character `'7'` is part of that string.
Pretty clever.
To see how filtering in list comprehensions translates to the list monad, we have to check out the `guard` function and the `Alternative` type class.

> The `Alternative` type class is for applicative functors that can also act as monoids.

Reviewer 2: ```suggestion
 The `Alternative` type class is for applicative functors that support a notion of failure (`empty`) and choice (`(<|>)`) that behave analogously to the operations of `Monoid`.
```

Here's its definition:

```{.haskell:hs}
class Applicative f => Alternative f where
    empty :: f a
    (<|>) :: f a -> f a -> f a
```

`empty` is synonymous to `mempty` from the `Monoid` type class and `(<|>)` corresponds to `<>`.
Because lists are monoids as well as monads, they can be made an instance of this type class:

> 

Reviewer 2: ```suggestion
`empty` is analogous to `mempty` from `Monoid`, and `(<|>)` resembles to `<>`, but they aren't the same thing as a `Monoid` instance in general.
 Lists are an instance of `Alternative`, where `empty` is `[]` and `(<|>)` concatenates lists:
```
